### PR TITLE
Improve plugin resolution (fixes pnpm compatibility)

### PR DESCRIPTION
**Pull request by adamhl8**
Continuation of #11248

### CODE REVIEW

File: tests/integration/__tests__/plugin-resolution.js [hunk 1]
```diff
@@ -93,6 +93,55 @@ describe("automatically loads '@prettier/plugin-*' from --plugin-search-dir (dif
...
+  runPrettier("plugins/pnpmVirtualStore/prettier@0.0.0", [
+    "file.txt",
+    "--parser=foobar",
+  ]).test({
+    stdout: "foobar+contents" + EOL,
+    stderr: "",
+    status: 0,
+    write: [],
+  });
+});
```
Comment (fisker) [inline tests/integration/__tests__/plugin-resolution.js:143]: I don't think these tests are correctly testing the added part. The plugin search logic always starts from Prettier cli, if we want test, we'll have to put Prettier into those dirs.

File: src/common/load-plugins.js [hunk 2]
```diff
@@ -113,6 +142,20 @@ async function findPluginsInNodeModules(nodeModulesDir) {
   return pluginPackageJsonPaths.map(path.dirname);
 }
 
+async function findPluginDirsInPnpmVirtualStore(searchDirectory) {
+  const pluginPackageJsonPaths = await fastGlob(
+    ["prettier-plugin-*", "@*/node_modules"],
```
Comment (fisker) [inline src/common/load-plugins.js:147]: Shouldn't plugins also stored like `prettier-plugin-foo@1.0.0`? I don't know how the `@*/plugin-*` stored, but are you sure these patterns correct?

File: tests/integration/__tests__/plugin-resolution.js (hunk 1 above)
Comment (adamhl8) [reply tests/integration/__tests__/plugin-resolution.js:143]: I could be misunderstanding the code, but I think in `run-prettier.js` the cwd gets set properly in `run()`. Either way, if I log `autoLoadDir`, it is set to the expected directory, so it does seem prettier is starting from the right directory.

File: src/common/load-plugins.js (hunk 2 above)
Comment (adamhl8) [reply src/common/load-plugins.js:147]: In the pnpm virtual store, the three possible types of plugins (`@prettier/plugin-*`, and `@<name>/prettier-plugin-*`, `prettier-plugin-*`) are stored like this:

```
.pnpm/
├── @prettier+plugin-foo@<version>/
│   └── node_modules/
│       └── @prettier/
│           └── plugin-foo/
├── @<name>+prettier-plugin-bar@<version>/
│   └── node_modules/
│       └── @<name>/
│           └── prettier-plugin-bar/
└── prettier-plugin-baz@<version>/
    └── node_modules/
        └── prettier-plugin-baz/
```

I don't think the `+package-name` part of the directory used to be there, but now that it is, I can just have the patterns be `["prettier-plugin-*", "@prettier+plugin-*", "@*+prettier-plugin-*"]`.

**Comment by adamhl8** [comment]
@fisker Any chance this will be included in next?

**Pull request closed without merging**